routes/webauthn.go

```diff
@@ -0,0 +1,377 @@
+			http.Error(w, invalidJSONMessage, http.StatusBadRequest)
+			return
+		}
+	}
+
+	options, sessionData, err := h.userSrvc.WebAuthnBeginRegistration(user, req.CredentialName)
+	if err != nil {
+		conf.Log().Request(r).Error("WebAuthn registration begin failed", "error", err)
+		http.Error(w, "failed to begin registration", http.StatusInternalServerError)
+		return
+	}
+
+	// Store session data in the response for the frontend to include in the finish request
+	response := map[string]interface{}{
+		"options":     options,
+		"sessionData": sessionData,
+	}
+
+	w.Header().Set(contentTypeHeader, applicationJSONType)
+	json.NewEncoder(w).Encode(response)
+}
+
+// PostRegisterFinish completes WebAuthn credential registration
+func (h *WebAuthnHandler) PostRegisterFinish(w http.ResponseWriter, r *http.Request) {
+	user := middlewares.GetPrincipal(r)
+	if user == nil {
+		conf.Log().Request(r).Error("WebAuthn registration finish - no user")
+		http.Error(w, "unauthorized", http.StatusUnauthorized)
+		return
+	}
+
+	conf.Log().Request(r).Info("WebAuthn registration finish started", "userID", user.ID)
+
+	// Read the raw body first for debugging
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "failed to read request body", http.StatusBadRequest)
+		return
+	}
+
+	var req struct {
+		SessionData interface{} `json:"sessionData"`
+		Response    interface{} `json:"response"`
+	}
+
+	if err := json.Unmarshal(body, &req); err != nil {
+		http.Error(w, invalidJSONMessage, http.StatusBadRequest)
+		return
+	}
+
+	conf.Log().Request(r).Info("WebAuthn registration finish - JSON decoded successfully")
+
+	// Convert response to JSON bytes for the service layer
+	responseBytes, err := json.Marshal(req.Response)
+	if err != nil {
+		conf.Log().Request(r).Error("WebAuthn registration finish - response marshal failed", "error", err)
+		http.Error(w, "invalid credential response", http.StatusBadRequest)
+		return
+	}
+
+	conf.Log().Request(r).Info("WebAuthn registration finish - calling service layer")
+
+	err = h.userSrvc.WebAuthnFinishRegistration(user, req.SessionData, responseBytes)
+	if err != nil {
+		conf.Log().Request(r).Error("WebAuthn registration finish failed", "error", err)
+		http.Error(w, "failed to finish registration", http.StatusBadRequest)
+		return
+	}
+
+	conf.Log().Request(r).Info("WebAuthn registration finish - success")
+
+	w.Header().Set(contentTypeHeader, applicationJSONType)
+	json.NewEncoder(w).Encode(map[string]interface{}{
+		"success": true,
+		"message": "WebAuthn credential registered successfully",
+	})
+}
+
+// PostLoginBegin starts WebAuthn authentication
+func (h *WebAuthnHandler) PostLoginBegin(w http.ResponseWriter, r *http.Request) {
+	var req struct {
+		Username string `json:"username,omitempty"`
+	}
+
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http.Error(w, invalidJSONMessage, http.StatusBadRequest)
+		return
+	}
+
+	// Support both username-based and usernameless flows
```

Comment:
What is the benefit of support username-based flow in addition? Couldn't we just support usernameless?